test/poll_for_state.go

```diff
@@ -16,41 +18,100 @@ type ChainAcker interface {
 // PollForAck attempts to find an acknowledgement containing a packet equal to the packet argument.
 // Polling starts at startHeight and continues until maxHeight. It is safe to call this function even if
 // the chain has yet to produce blocks for the target min/max height range. Polling delays until heights exist
-// on the chain. If no acknowledgement found, returns a not-found error.
-func PollForAck(ctx context.Context, chain ChainAcker, startHeight, maxHeight uint64, packet ibc.Packet) (zero ibc.PacketAcknowledgement, _ error) {
+// on the chain. Returns an error if acknowledgement not found or problems getting height or acknowledgements.
+func PollForAck(ctx context.Context, chain ChainAcker, startHeight, maxHeight uint64, packet ibc.Packet) (ibc.PacketAcknowledgement, error) {
+	poller := blockPoller{CurrentHeight: chain.Height, Acker: chain}
+	found, err := poller.doPoll(ctx, startHeight, maxHeight, packet)
+	if err != nil {
+		return ibc.PacketAcknowledgement{}, err
+	}
+	return found.(ibc.PacketAcknowledgement), nil
+}
+
+// ChainTimeouter is a chain that can get its timeouts at a specified height
+type ChainTimeouter interface {
+	ChainHeighter
+	Timeouts(ctx context.Context, height uint64) ([]ibc.PacketTimeout, error)
+}
+
+// PollForTimeout attempts to find a timeout containing a packet equal to the packet argument.
+// Otherwise, works identically to PollForAck.
+func PollForTimeout(ctx context.Context, chain ChainTimeouter, startHeight, maxHeight uint64, packet ibc.Packet) (ibc.PacketTimeout, error) {
+	poller := blockPoller{CurrentHeight: chain.Height, Timeouter: chain}
+	found, err := poller.doPoll(ctx, startHeight, maxHeight, packet)
+	if err != nil {
+		return ibc.PacketTimeout{}, err
+	}
+	return found.(ibc.PacketTimeout), nil
+}
+
+type blockPoller struct {
+	CurrentHeight func(ctx context.Context) (uint64, error)
+	Acker         ChainAcker
+	Timeouter     ChainTimeouter
+}
+
+func (p blockPoller) doPoll(ctx context.Context, startHeight, maxHeight uint64, packet ibc.Packet) (interface{}, error) {
 	if maxHeight < startHeight {
 		panic("maxHeight must be greater than or equal to startHeight")
 	}
 	var (
 		cursor  = startHeight
-		lastErr error
+		findErr error
+		found   interface{}
 	)
-
 	for cursor <= maxHeight {
-		curHeight, err := chain.Height(ctx)
+		curHeight, err := p.CurrentHeight(ctx)
 		if err != nil {
-			return zero, err
+			return nil, err
 		}
 		if cursor > curHeight {
 			continue
 		}
 
-		acks, err := chain.Acknowledgements(ctx, cursor)
-		if err != nil {
-			lastErr = err
+		switch {
+		case p.Acker != nil:
+			found, findErr = p.findAck(ctx, cursor, packet)
+		case p.Timeouter != nil:
+			found, findErr = p.findTimeout(ctx, cursor, packet)
+		default:
+			panic("poller misconfiguration")
+		}
```

Comment:
The switch could be expanded to reject `p.Acker != nil && p.Timeouter != nil`, but since this is an internal interface, it is unlikely that someone will set both... right?

Reply:
They better not! And that's what I was counting on. This type is a bit ugly but it's 100% private, so my expectation was a future developer should understand this and add unit tests around any new `poll` function.